### PR TITLE
CU-1jzahap: Policy premiums array refactoring

### PR DESCRIPTION
**Pull request by tgeorgas**
Premiums are now specified as follows:
[date1, premium1, date2, premium2, ...]

Removed premiumInterval. From now on the premiums can be specified at any time between the initiation and maturation dates.

calculateMaxNumOfPremiums() has been removed as it is no longer needed. Instead, the policy constructor checks to make sure the premiums are in increasing order and between the initiation date and the maturation date.

Updated unit tests to take permiumsDiff instead of premiums and tranchData
Diff instead of tranchData. These inputs include the premium input in the new format, but with the time difference from the initiation date rather than an absolute date.

Tranche is spelled "tranche" everywhere.

### CODE REVIEW

**Comment by melks** [comment]
Task linked: [CU-1jzahap Change premium variable in smart contracts to shorten its length and replace premium interval with an array of ordered timestamps. - BREAKING](https://app.clickup.com/t/1jzahap)

**Comment by melks** [comment]
Task linked: [CU-jf4uv5 Document Hash on Policy](https://app.clickup.com/t/jf4uv5)

File: test/utils/index.js [hunk 1]
```diff
@@ -173,7 +201,7 @@ export const createPolicy = async (entity, attrs = {}, ...callAttrs) => {
     startDate = currentTime + 120,
     maturationDate = currentTime + 300,
     unit = ADDRESS_ZERO,
-    premiumIntervalSeconds = 30,
+    // premiumIntervalSeconds = 30,
```
Comment (hiddentao) [inline test/utils/index.js:204]: pls remove

File: test/utils/index.js [hunk 2]
```diff
@@ -192,7 +220,7 @@ export const createPolicy = async (entity, attrs = {}, ...callAttrs) => {
     policyId,
     [
       type, 
-      premiumIntervalSeconds, 
+      // premiumIntervalSeconds,
```
Comment (hiddentao) [inline test/utils/index.js:223]: pls remove

File: test/utils/index.js [hunk 3]
```diff
@@ -237,7 +265,7 @@ export const preSetupPolicy = async (ctx, createPolicyArgs) => {
     startDate: currentBlockTime + startDateDiff,
     maturationDate: currentBlockTime + maturationDateDiff,
     unit: ctx.etherToken.address,
-    premiumIntervalSeconds,
+    // premiumIntervalSeconds,
```
Comment (hiddentao) [inline test/utils/index.js:268]: pls remove

File: contracts/PolicyCoreFacet.sol [hunk 4]
```diff
@@ -69,50 +68,71 @@ contract PolicyCoreFacet is EternalStorage, Controller, IDiamondFacet, IPolicyCo
...
-    uint256 nextPayTime = dataUint256["initiationDate"];
+    // uint256 nextPayTime = dataUint256["initiationDate"];
     uint256 numPremiums = 0;
-    for (uint256 p = 0; _premiums.length > p; p += 1) {
-      // we only care about premiums > 0
-      if (_premiums[p] > 0) {
-        dataUint256[__ii(i, numPremiums, "premiumAmount")] = _premiums[p];
-        dataUint256[__ii(i, numPremiums, "premiumDueAt")] = nextPayTime;
+
+    // for (uint256 p = 0; _premiums.length > p; p += 1) {
```
Comment (hiddentao) [inline contracts/PolicyCoreFacet.sol:97]: pls remove

File: contracts/PolicyCoreFacet.sol [hunk 5]
```diff
@@ -69,50 +68,71 @@ contract PolicyCoreFacet is EternalStorage, Controller, IDiamondFacet, IPolicyCo
...
+    // require(dataUint256["numTranches"] <= 99, 'max tranches reached');
 
-    // setup initial data for tranch
+    // setup initial data for tranche
     dataUint256[__i(i, "numShares")] = _numShares;
     dataUint256[__i(i, "pricePerShareAmount")] = _pricePerShareAmount;
 
     // iterate through premiums and figure out what needs to paid and when
-    uint256 nextPayTime = dataUint256["initiationDate"];
+    // uint256 nextPayTime = dataUint256["initiationDate"];
```
Comment (hiddentao) [inline contracts/PolicyCoreFacet.sol:94]: pls remove

File: contracts/PolicyCoreFacet.sol [hunk 6]
```diff
@@ -69,50 +68,71 @@ contract PolicyCoreFacet is EternalStorage, Controller, IDiamondFacet, IPolicyCo
...
   {
     require(_numShares > 0, 'invalid num of shares');
     require(_pricePerShareAmount > 0, 'invalid price');
 
-    // tranch count
+    // tranche count
     uint256 i = dataUint256["numTranches"];
     dataUint256["numTranches"] = i + 1;
-    require(dataUint256["numTranches"] <= 99, 'max tranches reached');
+    // require(dataUint256["numTranches"] <= 99, 'max tranches reached');
```
Comment (hiddentao) [inline contracts/PolicyCoreFacet.sol:87]: pls remove

File: contracts/PolicyCoreFacet.sol [hunk 7]
```diff
@@ -69,50 +68,71 @@ contract PolicyCoreFacet is EternalStorage, Controller, IDiamondFacet, IPolicyCo
...
+
+    uint256 previousPremiumDueAt = 0;
+    for (uint256 p = 0; _premiums.length > p; p += 2) {
+
+      if (_premiums[p+1] > 0) {
+
+        dataUint256[__ii(i, numPremiums, "premiumDueAt")] = _premiums[p];
+        dataUint256[__ii(i, numPremiums, "premiumAmount")] = _premiums[p+1];
+
+        require(dataUint256["initiationDate"] <= _premiums[p], 'premium befire initiation');
```
Comment (hiddentao) [inline contracts/PolicyCoreFacet.sol:118]: "before"

File: test/utils/index.js [hunk 8]
```diff
@@ -246,7 +274,8 @@ export const preSetupPolicy = async (ctx, createPolicyArgs) => {
     underwriter,
     claimsAdmin,
     insuredParty,
-    trancheData
+    // trancheData
```
Comment (hiddentao) [inline test/utils/index.js:277]: pls remove

File: test/utils/index.js [hunk 9]
```diff
@@ -136,21 +136,49 @@ export const outputBNs = bn => {
...
     ...callAttrs,
   )
 }
 
+export const adjustTrancheDataDiff = (policyInitiationDate, trancheDiffData) => {
+
+  let trancheData = trancheDiffData
+
+  if (trancheDiffData != undefined){
+  for (let i = 0; i < trancheDiffData.length; i += 1) {
```
Comment (hiddentao) [inline test/utils/index.js:169]: indentation

File: test/utils/index.js [hunk 10]
```diff
@@ -136,21 +136,49 @@ export const outputBNs = bn => {
...
+
+  return policy.createTranche(
     numShares,
     pricePerShareAmount,
     premiums,
     ...callAttrs,
   )
 }
 
+export const adjustTrancheDataDiff = (policyInitiationDate, trancheDiffData) => {
```
Comment (hiddentao) [inline test/utils/index.js:164]: does not need `export`

Comment (hiddentao) [reply test/utils/index.js:164]: Also, this logic seems to be duplicated within `createTranche()` - pls refactor

**Review by hiddentao — CHANGES_REQUESTED** [review]
See comments

File: contracts/PolicyCoreFacet.sol [hunk 11]
```diff
@@ -212,13 +232,13 @@ contract PolicyCoreFacet is EternalStorage, Controller, IDiamondFacet, IPolicyCo
     }
   }
 
-  function calculateMaxNumOfPremiums() public view override returns (uint256) {
-    return (dataUint256["maturationDate"] - dataUint256["initiationDate"]) / dataUint256["premiumIntervalSeconds"] + 1;
-  }
+  // function calculateMaxNumOfPremiums() public view override returns (uint256) {
```
Comment (hiddentao) [inline contracts/PolicyCoreFacet.sol:235]: pls remove

File: contracts/base/IPolicyCoreFacet.sol [hunk 12]
```diff
@@ -74,12 +73,13 @@ interface IPolicyCoreFacet {
...
-   *
-   * @return Max. no. of premium payments possible.
-   */
-  function calculateMaxNumOfPremiums() external view returns (uint256);
+  // /**
+  //  * @dev Get the max. no. of premium payments possible based on the policy dates.
+  //  *
+  //  * @return Max. no. of premium payments possible.
+  //  */
+  // function calculateMaxNumOfPremiums() external view returns (uint256);
```
Comment (hiddentao) [inline contracts/base/IPolicyCoreFacet.sol:81]: pls remove

File: contracts/test/DummyPolicyFacet.sol [hunk 13]
```diff
@@ -7,11 +7,13 @@ import '../base/IPolicyCoreFacet.sol';
 contract DummyPolicyFacet is IDiamondFacet, IPolicyCoreFacet {
   function getSelectors () public pure override returns (bytes memory) {
     return abi.encodePacked(
-      IPolicyCoreFacet.calculateMaxNumOfPremiums.selector
+      //Todo: why only this function here?
+      // IPolicyCoreFacet.calculateMaxNumOfPremiums.selector
```
Comment (hiddentao) [inline contracts/test/DummyPolicyFacet.sol:11]: pls remove

File: contracts/test/DummyPolicyFacet.sol [hunk 14]
```diff
@@ -44,9 +45,9 @@ contract DummyPolicyFacet is IDiamondFacet, IPolicyCoreFacet {
 
   function checkAndUpdateState () external override {}
 
-  function calculateMaxNumOfPremiums() external view override returns (uint256) {
-    return 666;
-  }
+  // function calculateMaxNumOfPremiums() external view override returns (uint256) {
```
Comment (hiddentao) [inline contracts/test/DummyPolicyFacet.sol:48]: pls remove

File: contracts/PolicyCoreFacet.sol [hunk 15]
```diff
@@ -149,13 +169,13 @@ contract PolicyCoreFacet is EternalStorage, Controller, IDiamondFacet, IPolicyCo
     startDate_ = dataUint256["startDate"];
     maturationDate_ = dataUint256["maturationDate"];
     unit_ = dataAddress["unit"];
-    premiumIntervalSeconds_ = dataUint256["premiumIntervalSeconds"];
+    // premiumIntervalSeconds_ = dataUint256["premiumIntervalSeconds"];
```
Comment (hiddentao) [inline contracts/PolicyCoreFacet.sol:172]: pls remove

File: contracts/PolicyCoreFacet.sol [hunk 16]
```diff
@@ -138,7 +158,7 @@ contract PolicyCoreFacet is EternalStorage, Controller, IDiamondFacet, IPolicyCo
     uint256 startDate_,
     uint256 maturationDate_,
     address unit_,
-    uint256 premiumIntervalSeconds_,
+    // uint256 premiumIntervalSeconds_,
```
Comment (hiddentao) [inline contracts/PolicyCoreFacet.sol:161]: pls remove

File: test/utils/index.js (hunk 1 above)
Comment (tgeorgas) [reply test/utils/index.js]: done

File: test/utils/index.js (hunk 2 above)
Comment (tgeorgas) [reply test/utils/index.js]: done

File: test/utils/index.js (hunk 3 above)
Comment (tgeorgas) [reply test/utils/index.js]: done

File: contracts/PolicyCoreFacet.sol (hunk 4 above)
Comment (tgeorgas) [reply contracts/PolicyCoreFacet.sol]: done

File: contracts/PolicyCoreFacet.sol (hunk 5 above)
Comment (tgeorgas) [reply contracts/PolicyCoreFacet.sol]: done

File: contracts/PolicyCoreFacet.sol (hunk 6 above)
Comment (tgeorgas) [reply contracts/PolicyCoreFacet.sol]: done

File: contracts/PolicyCoreFacet.sol (hunk 7 above)
Comment (tgeorgas) [reply contracts/PolicyCoreFacet.sol]: done

File: test/utils/index.js (hunk 8 above)
Comment (tgeorgas) [reply test/utils/index.js]: done

File: test/utils/index.js (hunk 9 above)
Comment (tgeorgas) [reply test/utils/index.js]: done

File: contracts/PolicyCoreFacet.sol (hunk 11 above)
Comment (tgeorgas) [reply contracts/PolicyCoreFacet.sol:217]: done

File: contracts/base/IPolicyCoreFacet.sol (hunk 12 above)
Comment (tgeorgas) [reply contracts/base/IPolicyCoreFacet.sol:81]: done

File: contracts/test/DummyPolicyFacet.sol (hunk 13 above)
Comment (tgeorgas) [reply contracts/test/DummyPolicyFacet.sol:11]: done

File: contracts/PolicyCoreFacet.sol (hunk 15 above)
Comment (tgeorgas) [reply contracts/PolicyCoreFacet.sol:154]: done

File: contracts/PolicyCoreFacet.sol (hunk 16 above)
Comment (tgeorgas) [reply contracts/PolicyCoreFacet.sol:143]: done

File: test/utils/index.js (hunk 10 above)
Comment (tgeorgas) [reply test/utils/index.js]: done.

File: contracts/test/DummyPolicyFacet.sol (hunk 14 above)
Comment (tgeorgas) [reply contracts/test/DummyPolicyFacet.sol]: done

**Review by hiddentao — APPROVED** [review]

File: contracts/PolicyCoreFacet.sol [hunk 17]
```diff
@@ -69,50 +68,53 @@ contract PolicyCoreFacet is EternalStorage, Controller, IDiamondFacet, IPolicyCo
...
+
+    uint256 previousPremiumDueAt = 0;
+    for (uint256 p = 0; _premiums.length > p; p += 2) {
+
+      if (_premiums[p+1] > 0) {
+
+        dataUint256[__ii(i, numPremiums, "premiumDueAt")] = _premiums[p];
+        dataUint256[__ii(i, numPremiums, "premiumAmount")] = _premiums[p+1];
+
+        require(dataUint256["initiationDate"] <= _premiums[p], 'premium before initiation');
```
Comment (hiddentao) [inline contracts/PolicyCoreFacet.sol:103]: Needs a test -> see https://coveralls.io/builds/44545758/source?filename=contracts%2FPolicyCoreFacet.sol

File: contracts/PolicyCoreFacet.sol [hunk 18]
```diff
@@ -69,50 +68,53 @@ contract PolicyCoreFacet is EternalStorage, Controller, IDiamondFacet, IPolicyCo
...
+        require(_premiums[p] <= dataUint256["maturationDate"], 'premium after maturation');
+        require(_premiums[p] > previousPremiumDueAt, 'premiums not in increasing order');
```
Comment (hiddentao) [inline contracts/PolicyCoreFacet.sol:105]: Needs a test -> see https://coveralls.io/builds/44545758/source?filename=contracts%2FPolicyCoreFacet.sol

**Review by hiddentao — CHANGES_REQUESTED** [review]
Need two more tests.

File: contracts/PolicyCoreFacet.sol (hunk 18 above)
Comment (tgeorgas) [reply contracts/PolicyCoreFacet.sol:105]: done

**Review by hiddentao — APPROVED** [review]
I like the addition of the message for chai assertions

**Pull request merged**